Expensify/App
# Setup push notifications

**Fixes:** https://github.com/Expensify/ReactNativeChat/issues/288

### Details

- Adds Airship as a library (`urbanairship-react-native`)
    - Airship's react native library is only compatible with iOS 11+, so I had to bump the iOS platform target
- Configures iOS and Android with Airship API keys, capabilities, permissions, etc...
- Refactors the Notification library to separate out LocalNotifications vs PushNotifications
    - LocalNotifications
        - Are triggered by the client.
        - Are currently only implemented for web/desktop using browser notifications. In the future, LocalNotifications could be implemented for mobile via a growl, snackbar, or using the native notification system (which I think could be done pretty easily using [wix/react-native-notifications](https://github.com/wix/react-native-notifications)
    - PushNotifications
        - Are triggered remotely by the Airship API.
        - By default, are only displayed if the app is in the background.
        - Are only implemented for native platforms, and don't really have an equivalent on web/desktop.
        - Furthermore, if the app is in the foreground, any callbacks associated with push notifications will not be executed. We're assuming that if we're receiving a push notification, then we must be connected to the internet, and if we're connected to the internet, then we must be connected to Pusher. And if we're connected to pusher, then the events having to do with Push notifications will be caught and handled via pusher.
- Gets permission to display push notifications when the app first loads
    - If user has already granted or denied permissions for user notifications, this code not affect the user at all.
    - From the Airship docs:

        > The Airship module makes a distinction between 'user notifications,' which can be seen by the user, and 'invisible notifications' that carry only data for the app to process.
    - Even if the user denies permission for 'user notifications', we will still be able to send push notifications to trigger update actions in the background.
- On sign-in, registers the device for push notifications (associates the device and accountID with a named_user in the Airship API) On sign-out, deregisters the device for push notifications (associates the device with an undefined named_user in the Airship API)
    - To implement this, we simply subscribe the push notification library to `IONKEYS.SESSION.accountID`. When the user signs in, the `SESSION.accountID` is assigned, and the registration is triggered. When the user signs out, we clear Ion, the `SESSION.accountID` is nullified, and the deregistration is triggered. However, we weren't executing the callbacks for Ion subscribers when we called `Ion.clear()`, so this PR implements that as well.

### Tests / QA

1. Build the mobile app to a physical iPhone device (instructions [here](https://stackoverflow.com/c/expensify/questions/6735)). (dev) Or just open the iPhone app 😉 (QA)
2. Log into an account and background the app.
3.  Go to `expensify.com.dev` as a different user, and submit an expense report to the account you used in step 2.
4. Comment on the report, and verify that you receive a push notification on your device with the comment information.
5. Open the app, and verify that you see the report comment from step 4.
6. Sign into an account.
7. Run `npm run desktop`, and use the desktop UI to sign into another account.
8. Send a chat from the desktop account to the account you used on mobile, with the app open on mobile. Ensure that you _do not_ see a push notification for the chat.
9. Send a chat from the desktop account to the account you used on mobile, with the app backgrounded on mobile. Ensure that you see a push notification.
10. Close the app on mobile. (i.e: actually close it by swiping up and then swiping it away, don't just background it)
11. Send a chat from the desktop account to the account you used on mobile. Ensure that you see a push notification.
12. Click on the push notification, and see that it opens up the app to the correct chat.
13. (dev) Find the accountID of the account you signed in on in mobile, and find a report ID for any of its reports. Then run `$EXPENSIDEV/script/clitools push:reportComment`, follow the prompts, and verify that you see the push notification show up.
14. Log out on mobile.
15. Send a chat from the web account to the account you used on mobile. Ensure that you don't see a push notification.
16. Build the mobile app to an Android emulator, and repeat steps 2-15.
